server/pbenchinacan/run-pbench-in-a-can

```diff
@@ -1,24 +1,74 @@
 #!/bin/bash
 
+# The default container registry to use, and the organization name in that
+# registry.
 export PB_CONTAINER_REG=${PB_CONTAINER_REG:-"images.paas.redhat.com"}
 export PB_ORG_NAME=${PB_ORG_NAME:-"pbench"}
 
+# Typically, the CI builds the Pbench Server container image and pushes it to
+# the container registry as part of the CI pipeline.  However, for development,
+# we need to build a local Pbench Server RPM and image from the checked-out
+# sources and run the functional tests against it.  But, we don't want to
+# require pushing that resulting image to the container registry in order to
+# pull it back to run it in the pod below.  To enable this, we use a separate
+# environment variable just for the Pbench Server.
+export PB_SERVER_CONTAINER_REG=${PB_SERVER_CONTAINER_REG:-${PB_CONTAINER_REG}}
+
+# Default pull policy for all images is to always try to pull the image (NOTE:
+# this does not mean it will download the image, but it will check to see if the
+# exact copy is available locally).
+export PB_COMMON_IMAGE_PULL_POLICY=${PB_COMMON_IMAGE_PULL_POLICY:-"Always"}
+
+# Each container has its own name and tag for the image, as well as a pull
+# policy.
 export PB_ELASTICSEARCH_IMAGE_NAME=${PB_ELASTICSEARCH_IMAGE_NAME:-"elasticsearch"}
 export PB_ELASTICSEARCH_IMAGE_TAG=${PB_ELASTICSEARCH_IMAGE_TAG:-"7.13.3"}
+if [[ "${PB_CONTAINER_REG}" == "localhost" ]]; then
+    # Drop the explict registry reference when using "localhost".
+    export PB_ELASTICSEARCH_IMAGE="${PB_ELASTICSEARCH_IMAGE_NAME}:${PB_ELASTICSEARCH_IMAGE_TAG}"
+    export PB_ELASTICSEARCH_IMAGE_PULL_POLICY="Never"
+else
+    export PB_ELASTICSEARCH_IMAGE="${PB_CONTAINER_REG}/${PB_ORG_NAME}/${PB_ELASTICSEARCH_IMAGE_NAME}:${PB_ELASTICSEARCH_IMAGE_TAG}"
+    export PB_ELASTICSEARCH_IMAGE_PULL_POLICY="${PB_ELASTICSEARCH_IMAGE_PULL_POLICY:-${PB_COMMON_IMAGE_PULL_POLICY}}"
+fi
+
 export PB_POSTGRESQL_IMAGE_NAME=${PB_POSTGRESQL_IMAGE_NAME:-"postgresql-13"}
 export PB_POSTGRESQL_IMAGE_TAG=${PB_POSTGRESQL_IMAGE_TAG:-"latest"}
+if [[ "${PB_CONTAINER_REG}" == "localhost" ]]; then
+    # Drop the explict registry reference when using "localhost".
+    export PB_POSTGRESQL_IMAGE="${PB_POSTGRESQL_IMAGE_NAME}:${PB_POSTGRESQL_IMAGE_TAG}"
```

Comment:
So what happens if `localhost` doesn't already have Elasticsearch or Keycloak or PostgreSQL? Don't we need the registry to pull them??? And given that we're only pulling if we *don't* have them, what's the advantage in dropping the registry?

(Keycloak might be an exception only if we have similar local-only build logic for our beefed-up Keycloak image, which I don't think plays in this sandbox currently...)

Reply:
> So what happens if `localhost` doesn't already have Elasticsearch or Keycloak or PostgreSQL?

It will fail to find the image just as it does before this change when you `export PB_CONTAINER_REG=localhost`.  I shouldn't be dropping the `localhost`, but we have to stop using "pull policy" when we do.  I'll fix the `localhost` reference.

> Don't we need the registry to pull them???

Yes, but if somebody explicitly sets `PB_CONTAINER_REG=localhost` they are communicating that they know what they are doing.

> And given that we're only pulling if we _don't_ have them, what's the advantage in dropping the registry?

As above, if you have an image reference of `localhost/pbench-server-ci:main`, for example, you can't set a pull policy in the pod specification to `Always` pull the image, the pod will fail to start.  So we set it to never, since `localhost` is being used.

> (Keycloak might be an exception only if we have similar local-only build logic for our beefed-up Keycloak image, which I don't think plays in this sandbox currently...)

Agreed.

Reply:
Fixed.